src/main/java/org/icgc/argo/program_service/services/ego/EgoService.java

```diff
@@ -0,0 +1,253 @@
+import lombok.val;
+import org.icgc.argo.program_service.proto.User;
+import org.icgc.argo.program_service.proto.UserRole;
+import org.icgc.argo.program_service.converter.ProgramConverter;
+import org.icgc.argo.program_service.model.entity.ProgramEgoGroupEntity;
+import org.icgc.argo.program_service.model.entity.ProgramEntity;
+import org.icgc.argo.program_service.repositories.ProgramEgoGroupRepository;
+import org.icgc.argo.program_service.services.ego.model.entity.*;
+import org.icgc.argo.program_service.services.ego.model.exceptions.EgoException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.HttpServerErrorException;
+
+import javax.transaction.Transactional;
+import javax.validation.constraints.Email;
+import java.security.interfaces.RSAPublicKey;
+import java.util.*;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static org.icgc.argo.program_service.proto.UserRole.ADMIN;
+import static org.icgc.argo.program_service.services.ego.GroupName.createProgramGroupName;
+
+//TODO [rtisma]: refactor into service and client
+@Slf4j
+@Service
+public class EgoService {
+  private final EgoRESTClient egoClient;
+  private final ProgramEgoGroupRepository programEgoGroupRepository;
+  private final ProgramConverter programConverter;
+  private RSAPublicKey egoPublicKey;
+
+  @Autowired
+  public EgoService(
+    @NonNull ProgramEgoGroupRepository programEgoGroupRepository,
+	  @NonNull ProgramConverter programConverter,
+    @NonNull EgoRESTClient restClient
+      ) {
+    this.programEgoGroupRepository = programEgoGroupRepository;
+    this.programConverter = programConverter;
+    this.egoClient = restClient;
+  }
+
+  @Autowired
+  private void setEgoPublicKey() {
+    this.egoPublicKey = egoClient.getPublicKey();
+  }
+
+
+  public Optional<EgoToken> verifyToken(String jwtToken) {
+    try {
+      Algorithm algorithm = Algorithm.RSA256(this.egoPublicKey, null);
+      JWTVerifier verifier = JWT.require(algorithm)
+        .withIssuer("ego")
+        .build(); //Reusable verifier instance
+      val jwt = verifier.verify(jwtToken);
+      return parseToken(jwt);
+    } catch (JWTVerificationException | NullPointerException e) {
+      // Handle NPE defensively for null JWT.
+      return Optional.empty();
+    }
+  }
+
+  public Optional<EgoToken> parseToken(DecodedJWT jwt) {
+    try {
+      EgoToken egoToken = new EgoToken(jwt, jwt.getClaim("context").as(Context.class));
+      return Optional.of(egoToken);
+    } catch (JWTDecodeException exception){
+      //Invalid token
+      return Optional.empty();
+    }
+  }
+
+
+  //TODO: add transactional. If there are more programdb logic in the future and something fails, it will be able to roll back those changes.
```

Comment:
:+1:

Reply:
Thanks for identifying this. This may be our first example for having to have a transaction across multiple microservices.

Reply:
Thank Rob for that one. :-) He identified it, not me. :-)